### PR TITLE
Changelog entries for 11.3.0

### PR DESCRIPTION
**Pull request by hanefi**
In this release, I tried something different. I experimented with adding the PR number and title to the changelog right before each changelog entry. This way, it is easier to track where a particular changelog entry comes from. After reviews are over, I plan to remove those lines with PR numbers and titles.

I went through all the PRs that are merged after 11.2.0 release and came up with a list of PRs that may need help with changelog entries. You can see details on PRs grouped in several sections below.

## PRs with missing entries

The following PRs below do not have a changelog entry. If you think that this is a mistake, please share it in this PR along with a suggestion on what the changelog item should be.

PR #6846 : fix 3 flaky tests in failure schedule
PR #6844 : Add CPU usage to citus_stat_tenants
PR #6833 : Fix citus_stat_tenants period updating bug
PR #6787 : Add more tests for ddl coverage
PR #6842 : Add build-cdc-* temporary directories to .gitignore
PR #6841 : Add build-cdc-* temporary directories to .gitignore
PR #6840 : Bump Citus to 12.0devel
PR #6824 : Fixes flakiness in multi_metadata_sync test
PR #6811 : Backport identity column improvements to v11.2
PR #6830 : In run_test.py actually return worker_count
PR #6825 : Fixes flakiness in multi_cluster_management test
PR #6816 : Refactor run_test.py
PR #6817 : Explicitly disallow local rels when inserting into dist table
PR #6821 : Rename citus stats tenants
PR #6822 : Add some more tests for initial sql support
PR #6819 : Fix flakyness in citus_split_shard_by_split_points_deferred_drop
PR #6814 : Make python-regress based tests runnable with run_test.py
PR #6813 : Fix flaky multi_mx_schema_support test
PR #6720 : Convert columnar tap tests to pytest
PR #6812 : Revoke statistics permissions from public and grant them to pg_monitor
PR #6769 : Citus stats tenants guc
PR #6807 : Fix the incorrect (constant) value passed to pointer-to-bool parameter, pass a NULL  as the value is not used
PR #6797 : Attribute local queries and cached plans on local execution
PR #6796 : Parse the annotation string correctly
PR #6762 : Add logs to citus_stats_tenants
PR #6773 : Add initial sql support for distributed tables that don't have a shard key
PR #6792 : Disentangle MERGE planning code from the modify-planning code path
PR #6761 : Citus stats tenants collector view
PR #6791 : Make 8 more tests runnable multiple times via run_test.py
PR #6786 : Refactor some of the planning code to accommodate a new planning path for MERGE SQL
PR #6789 : Rename AllRelations.. functions to AllDistributedRelations..
PR #6788 : Actually skip arbitrary_configs_router & nested_execution for AllNullDistKeyDefaultConfig
PR #6783 : Add a config for arbitrary config tests where all the tables are null-shard-key tables
PR #6784 : Fix attach partition: citus local to null distributed
PR #6782 :  Add an arbitrary config test heavily based on multi_router_planner_fast_path.sql
PR #6781 : Decide what to do with router planner error at one place
PR #6778 : Support partitioning for dist tables with null dist keys
PR #6766 : fix pip lock file
PR #6764 : Make workerCount configurable for regression tests
PR #6745 : Add support for creating distributed tables with a null shard key
PR #6696 : This implements MERGE phase-III
PR #6767 : Add pytest depedencies to Pipfile
PR #6760 : Decide core distribution params in CreateCitusTable
PR #6759 : Add multi_create_fdw into minimal_schedule
PR #6743 : Replace CITUS_TABLE_WITH_NO_DIST_KEY checks with HasDistributionKey()
PR #6751 : Stabilize single_node.sql and others that report illegal node removal
PR #6742 : Refactor CreateDistributedTable()
PR #6747 : Remove unused lock functions
PR #6744 : Fix multiple output version arbitrary config tests
PR #6741 : Stabilize single node tests
PR #6740 : Fix string eval bug in migration files check
PR #6736 : Make run_test.py and create_test.py importable without errors
PR #6734 : Don't blanket ignore flake8 E402 error
PR #6737 : Fixes bookworm packaging pipeline problem
PR #6735 : Fix run_test.py on python 3.9
PR #6733 : MERGE: In deparser, add missing check for RETURNING clause.
PR #6714 : Remove auto_explain workaround in citus explain hook for ALTER TABLE
PR #6719 : Fix flaky test
PR #6718 : Add more powerfull dependency tracking to run_test.py
PR #6710 : Install non-vulnerable cryptography package
PR #6711 : Support compilation and run tests on latest PG versions
PR #6700 : Add auto-formatting and linting to our python code
PR #6707 : Allow multi_insert_select to run repeatably
PR #6708 : Fix flakyness in failure_create_distributed_table_non_empty
PR #6698 : Miscellaneous cleanup
PR #6704 : Update README for 11.2
PR #6703 : Fix dubious ownership error from git
PR #6690 : Bump Citus to 11.3devel

## Too long changelog entries

The following PRs have changelog entries that are too long to fit in a single line. I'd expect authors to supply at changelog entries in `DESCRIPTION:` lines that are at most 78 characters. If you want to supply multi-line changelog items, you can have multiple lines that start with `DESCRIPTION:` instead.

PR #6837 : fixes update propagation bug when `citus_set_coordinator_host` is called more than once
PR #6738 :  Identity column implementation refactorings
PR #6756 : Schedule parallel shard moves in background rebalancer by removing task dependencies between shard moves across colocation groups.
PR #6793 : Add a GUC to disallow planning the queries that reference non-colocated tables via router planner
PR #6726 : fix memory leak during altering distributed table with a lot of partition and shards
PR #6722 : fix memory leak during distribution of a table with a lot of partitions
PR #6693 : prevent memory leak during ConvertTable with a lot of partitions

## Empty changelog entries.

The following PR had an empty `DESCRIPTION:` line. This generates an empty changelog line that needs to be removed manually. Please either provide a short entry, or remove `DESCRIPTION:` line completely.

PR #6810 : Make CDC decoder an independent extension
PR #6827 : Makefile changes to build CDC in builddir for pgoutput and wal2json.

### CODE REVIEW

**Comment by hanefi** [comment]
The following PRs do not have a changelog entry

PR #6846 : fix 3 flaky tests in failure schedule
PR #6844 : Add CPU usage to citus_stat_tenants
PR #6833 : Fix citus_stat_tenants period updating bug
PR #6787 : Add more tests for ddl coverage
PR #6842 : Add build-cdc-* temporary directories to .gitignore
PR #6841 : Add build-cdc-* temporary directories to .gitignore
PR #6840 : Bump Citus to 12.0devel
PR #6824 : Fixes flakiness in multi_metadata_sync test
PR #6811 : Backport identity column improvements to v11.2
PR #6830 : In run_test.py actually return worker_count
PR #6825 : Fixes flakiness in multi_cluster_management test
PR #6816 : Refactor run_test.py
PR #6817 : Explicitly disallow local rels when inserting into dist table
PR #6821 : Rename citus stats tenants
PR #6822 : Add some more tests for initial sql support
PR #6819 : Fix flakyness in citus_split_shard_by_split_points_deferred_drop
PR #6814 : Make python-regress based tests runnable with run_test.py
PR #6813 : Fix flaky multi_mx_schema_support test
PR #6720 : Convert columnar tap tests to pytest
PR #6812 : Revoke statistics permissions from public and grant them to pg_monitor
PR #6769 : Citus stats tenants guc
PR #6807 : Fix the incorrect (constant) value passed to pointer-to-bool parameter, pass a NULL  as the value is not used
PR #6797 : Attribute local queries and cached plans on local execution
PR #6796 : Parse the annotation string correctly
PR #6762 : Add logs to citus_stats_tenants
PR #6773 : Add initial sql support for distributed tables that don't have a shard key
PR #6792 : Disentangle MERGE planning code from the modify-planning code path
PR #6761 : Citus stats tenants collector view
PR #6791 : Make 8 more tests runnable multiple times via run_test.py
PR #6786 : Refactor some of the planning code to accommodate a new planning path for MERGE SQL
PR #6789 : Rename AllRelations.. functions to AllDistributedRelations..
PR #6788 : Actually skip arbitrary_configs_router & nested_execution for AllNullDistKeyDefaultConfig
PR #6783 : Add a config for arbitrary config tests where all the tables are null-shard-key tables
PR #6784 : Fix attach partition: citus local to null distributed
PR #6782 :  Add an arbitrary config test heavily based on multi_router_planner_fast_path.sql
PR #6781 : Decide what to do with router planner error at one place
PR #6778 : Support partitioning for dist tables with null dist keys
PR #6766 : fix pip lock file
PR #6764 : Make workerCount configurable for regression tests
PR #6745 : Add support for creating distributed tables with a null shard key
PR #6696 : This implements MERGE phase-III
PR #6767 : Add pytest depedencies to Pipfile
PR #6760 : Decide core distribution params in CreateCitusTable
PR #6759 : Add multi_create_fdw into minimal_schedule
PR #6743 : Replace CITUS_TABLE_WITH_NO_DIST_KEY checks with HasDistributionKey()
PR #6751 : Stabilize single_node.sql and others that report illegal node removal
PR #6742 : Refactor CreateDistributedTable()
PR #6747 : Remove unused lock functions
PR #6744 : Fix multiple output version arbitrary config tests
PR #6741 : Stabilize single node tests
PR #6740 : Fix string eval bug in migration files check
PR #6736 : Make run_test.py and create_test.py importable without errors
PR #6734 : Don't blanket ignore flake8 E402 error
PR #6737 : Fixes bookworm packaging pipeline problem
PR #6735 : Fix run_test.py on python 3.9
PR #6733 : MERGE: In deparser, add missing check for RETURNING clause.
PR #6714 : Remove auto_explain workaround in citus explain hook for ALTER TABLE
PR #6719 : Fix flaky test
PR #6718 : Add more powerfull dependency tracking to run_test.py
PR #6710 : Install non-vulnerable cryptography package
PR #6711 : Support compilation and run tests on latest PG versions
PR #6700 : Add auto-formatting and linting to our python code
PR #6707 : Allow multi_insert_select to run repeatably
PR #6708 : Fix flakyness in failure_create_distributed_table_non_empty
PR #6698 : Miscellaneous cleanup
PR #6704 : Update README for 11.2
PR #6703 : Fix dubious ownership error from git
PR #6690 : Bump Citus to 11.3devel


The following PRs have changelog entries that are too long to fit in a single line.

PR #6837 : fixes update propagation bug when `citus_set_coordinator_host` is called more than once
PR #6738 :  Identity column implementation refactorings
PR #6756 : Schedule parallel shard moves in background rebalancer by removing task dependencies between shard moves across colocation groups.
PR #6793 : Add a GUC to disallow planning the queries that reference non-colocated tables via router planner
PR #6726 : fix memory leak during altering distributed table with a lot of partition and shards
PR #6722 : fix memory leak during distribution of a table with a lot of partitions
PR #6693 : prevent memory leak during ConvertTable with a lot of partitions

The following PR had an empty `DESCRIPTION:` line.

PR #6810 : Make CDC decoder an independent extension

**Comment by pinodeca** [comment]
I like having the PR lnk in the change log, makes it so easy to click through for clarity or when intrigued.

File: CHANGELOG.md [hunk 1]
```diff
@@ -1,3 +1,103 @@
...
+  than once.
+
+PR #6771 : Adds control for background task executors involving a node
+* Adds control for background task executors involving a node
+
+PR #6820 : Add upgrade/downgrade paths between v11.2.2 and v11.3.1
+* PR description that will go into the change log, up to 78 characters
+
+PR #6794 : Move cleanup record test to upgrade schedule
+* Move cleanup record test to upgrade schedule
```
Comment (onurctirtir) [inline CHANGELOG.md:19]: both need to be removed

File: CHANGELOG.md [hunk 2]
```diff
@@ -1,3 +1,103 @@
...
+  background task monitor can run them concurrently.
+
+PR #6776 : Propagate CREATE/ALTER/DROP PUBLICATION statements
+* Propagate CREATE/ALTER/DROP PUBLICATION statements for distributed tables
+
+PR #6802 : Handle identity columns properly in the router planner
+* Fixes a bug with insert..select queries with identity columns
+
+PR #6775 : Add appropriate locks for MERGE to run in in parallel
+* PR description that will go into the change log, up to 78 characters
```
Comment (onurctirtir) [inline CHANGELOG.md:58]: does that fix a bug that we've introduced in this release or not?

Comment (onurctirtir) [reply CHANGELOG.md:58]: cc: @tejeswarm

File: CHANGELOG.md [hunk 3]
```diff
@@ -1,3 +1,103 @@
...
+
+PR #6805 : Do not override combinefunc of custom aggregates with common names
+* Fix an issue that caused some queries with custom aggregates to fail
+
+PR #6728 : stabilize metadata syncing
+* Fixes memory leak and max allocation block errors during metadata syncing.
+
+* Introduces nontransactional mode for metadata sync.
+
+* Introduces the GUC `citus.metadata_sync_mode` to switch sync modes
```
[the remark this answers is not in the document]
Comment (onurctirtir) [reply CHANGELOG.md:32]: cc: @aykut-bozkurt

File: CHANGELOG.md [hunk 4]
```diff
@@ -1,3 +1,103 @@
...
+* PR description that will go into the change log, up to 78 characters
+
+PR #6793 : Add a GUC to disallow planning the queries that reference
+non-colocated tables via router planner
+* Adds `citus.enable_non_colocated_router_query_pushdown` GUC to ensure
+  generating a consistent distributed plan for the queries that reference
+  non-colocated distributed tables (when set to "false", the default is "true").
+
+PR #6754 : Check before logicalrep for rebalancer, error if needed
+* Check before logicalrep for rebalancer, error if needed
```
Comment (onurctirtir) [inline CHANGELOG.md:67]: what do we "check" there? could we use a better wording? @agedemenli

File: CHANGELOG.md [hunk 5]
```diff
@@ -1,3 +1,103 @@
...
+
+PR #6721 : Exclude-Generated-Columns-In-Copy
+* Fixes a bug in shard copy operations.
+
+PR #6739 : Rebalance shard groups with placement count less than worker count
+* Adds logic to distribute unbalanced shards
```
Comment (onurctirtir) [inline CHANGELOG.md:73]: "Adds logic" sounds a bit off, should we instead tell:

Fixes a bug that prevents rebalancing shard groups with placement count less ..

File: CHANGELOG.md [hunk 6]
```diff
@@ -1,3 +1,103 @@
...
+PR #6748 : Use pg_total_relation_size in citus_shards
+* Correctly report shard size in citus_shards vie
+
+PR #6726 : fix memory leak during altering distributed table with a lot of
+partition and shards
+* Fixes a memory leak issue during altering distributed table with a lot of
+  partition and shards.
+
+PR #6732 : Move common code to a new routine.
+* PR description that will go into the change log, up to 78 characters
```
Comment (onurctirtir) [inline CHANGELOG.md:84]: need to remove I guess

File: CHANGELOG.md [hunk 7]
```diff
@@ -1,3 +1,103 @@
...
+
+PR #6724 : fix single tuple result memory leak
+* Fixes memory leak issue with query results that returns single row.
+
+PR #6693 : prevent memory leak during ConvertTable with a lot of partitions
+* Fixes memory leak during undistribution and alteration of a table with a lot
+  of partitions.
+
+PR #6688 : Stop background daemon before dropping the database
+* Stop maintenance daemon when dropping a database even without Citus extension
```
Comment (onurctirtir) [inline CHANGELOG.md:99]: makes sure to stop ..

File: CHANGELOG.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (aykut-bozkurt) [reply CHANGELOG.md:32]: Maybe repharesing like `Introduces nontransactional mode for metadata sync via the GUC citus.metadata_sync_mode`

**Comment by onurctirtir** [comment]
> PRs with missing entries
>
> The following PRs below do not have a changelog entry. If you think that this is a mistake, please share it in this PR along with a suggestion on what the changelog item should be.

None of those PRs require changelog entry except the `Merge Phase-III` one. We should definitely have a more descriptive entry for that.

**Comment by rajeshkt78** [comment]
I don't see the Changelog for CDC (Change Data Capture) PR in the Changelog..
https://github.com/citusdata/citus/pull/6623/
https://github.com/citusdata/citus/pull/6810
https://github.com/citusdata/citus/pull/6827

**Review by marcocitus — COMMENTED** [review]
Leaving a unrelated/semi-related note here: We probably need to bump the citus_columnar version as well even if there were no changes.

File: CHANGELOG.md (hunk 5 above)
Comment (hanefi) [reply CHANGELOG.md]: I do not think that this is a bug fix. It looks more like an improvement.

How about the following @agedemenli ?

```suggestion
* Improves rebalancer when shard groups have placement count less than worker
  count
```

**Comment by hanefi** [comment]
> I don't see the Changelog for CDC (Change Data Capture) PR in the Changelog.. #6623 #6810 #6827

We expect the changelog entry to be in the same line that starts with `DESCRIPTION: `. However #6810 and #6827 does not have that information on the same line. However, I do not really understand why #6623 was skipped. I guess we should have that.

The pull request template is `DESCRIPTION: PR description that will go into the change log, up to 78 characters`. Maybe it is not clear that we want the description on the same line.

How about a single line at the very beginning:

`* Introduces CDC implementation for Citus using logical replication (#6623, #6810, #6827)`

File: CHANGELOG.md (hunk 2 above)
Comment (tejeswarm) [reply CHANGELOG.md]: Done.

**Comment by aykut-bozkurt** [comment]
I fixed following too long entries:
#6837, #6726, #6722, #6693

**Comment by hanefi** [comment]
> PR https://github.com/citusdata/citus/pull/6733 : MERGE: In deparser, add missing check for RETURNING clause.
Added

The added changelog entry was: `RETURNING is not supported in MERGE, so, it must be NULL`

@tejeswarm I do not think that this one deserves an entry in the changelog as there is no supported MERGE command with returning clause as of today and it is impossible right now to get the error message.

Thanks for adding a nice entry for #6696 though. Appreciated.

File: CHANGELOG.md [hunk 8]
```diff
@@ -1,3 +1,70 @@
...
+  of background task executors involving a node (#6771)
+
+* Adds the view `citus_stats_tenants` that monitor statistics on tenant usages
+  (#6725)
+
+* Allows parallel shard moves in background rebalancer (#6756)
+
+* Checks before logicalrep for rebalancer, error if needed (#6754)
+
+* Correctly reports shard size in citus_shards view (#6748)
```
Comment (hanefi) [inline CHANGELOG.md:23]:
```suggestion
* Correctly reports shard size in `citus_shards` view (#6748)
```

File: CHANGELOG.md [hunk 9]
```diff
@@ -1,3 +1,103 @@
...
+set a user defined comparison function.
+* Fixes an uninitialized memory access in shard split API
+
+PR #6837 : fixes update propagation bug when `citus_set_coordinator_host` is
+called more than once
+* Fixes update propagation bug when `citus_set_coordinator_host` is called more
+  than once.
+
+PR #6771 : Adds control for background task executors involving a node
+* Adds control for background task executors involving a node
```
[the remark this answers is not in the document]
Comment (naisila) [reply CHANGELOG.md]: Sounds good

File: CHANGELOG.md [hunk 10]
```diff
@@ -1,3 +1,103 @@
...
+
+PR #6820 : Add upgrade/downgrade paths between v11.2.2 and v11.3.1
+* PR description that will go into the change log, up to 78 characters
+
+PR #6794 : Move cleanup record test to upgrade schedule
+* Move cleanup record test to upgrade schedule
+
+PR #6725 : Multi tenant monitoring
+* Adds views that monitor statistics on tenant usages
```
[the remark this answers is not in the document]
Comment (halilozanakgul) [reply CHANGELOG.md]: Looks better. Maybe
```
Adds `citus_stat_tenants` view that monitors statistics on tenants usages
```

There is actually one more view, `citus_stat_tenants_local` but it is not as important to mention I guess.

**Comment by onurctirtir** [comment]
Is this ready for reviews?


Also, as Marco mentioned, should we bump columnar version too?

> Leaving a unrelated/semi-related note here: We probably need to bump the citus_columnar version as well even if there were no changes.

File: CHANGELOG.md (hunk 10 above)
[the remark this answers is not in the document]
Comment (hanefi) [reply CHANGELOG.md]: I like to have a consistent structure with `Adds the <type> <object_name> that does ... ` here that results in better lexicographical ordering

**Comment by hanefi** [comment]
> Is this ready for reviews?

It is ready now.


> Also, as Marco mentioned, should we bump columnar version too?

I can (and will) do that in another PR.

**Comment by onurctirtir** [comment]
> > Also, as Marco mentioned, should we bump columnar version too?
> 
> I can (and will) do that in another PR.

We should probably decide the procedure for that:
* Should we bump schema version on minor schema versions too (-1, -2 ..) or only on major updates?
* Should we reflect this in release automation scripts? etc.

File: CHANGELOG.md [hunk 11]
```diff
@@ -1,3 +1,70 @@
+### citus v11.3.0 (April 13, 2023) ###
+
+* Introduces CDC implementation for Citus using logical replication
+  (#6623, #6810, #6827)
+
+* Adds support for `MERGE` command on co-located distributed tables joined on
+  distribution column (#6799)
+
+* Adds the GUC `citus.enable_non_colocated_router_query_pushdown` to ensure
```
[the remark this answers is not in the document]
Comment (onurctirtir) [reply CHANGELOG.md:9]: I see what Onder means, let's move this to right after
> Allows parallel shard moves in background rebalancer (#6756)

[the remark this answers is not in the document]
Comment (onurctirtir) [reply CHANGELOG.md:9]: I will push a commit to a separate branch to suggest a different ordering for the entries

[the remark this answers is not in the document]
Comment (hanefi) [reply CHANGELOG.md]: I forgot to cherry pick your comment Onur. Now I did it and the changelog should be ok.

**Comment by hanefi** [comment]
A final update:

- I added one more changelog entry for a newly backported commit in 8650101
- I cherry picked Onur's commit that reorders some items as suggested by Onder. 5f03124
- I opened 2 PRs that update columnar versions on `release-11.3` and `main` branches. #6897 and #6898
- I see 3 failures on packaging tests that are not really related to my changes in the PR. I hope the connection timeouts for the PGDG repositories get resolved soon and the tests start to pass again.

File: CHANGELOG.md [hunk 12]
```diff
@@ -1,3 +1,73 @@
...
+  non-colocated distributed tables when set to "false" (#6793)
+
+* Checks if all moves are able to be done via logical replication for rebalancer
+  (#6754)
+
+* Correctly reports shard size in `citus_shards` view (#6748)
+
+* Fixes a bug in shard copy operations (#6721)
+
+* Fixes a bug related to sequence idempotency at non-transactional sync (#6889)
```
Comment (onurctirtir) [inline CHANGELOG.md:34]: We don't need to have this for 12.0 because non-transactional metadata sync is introduced in this release already?

**Review by onurctirtir — APPROVED** [review]

File: CHANGELOG.md (hunk 12 above)
Comment (onurctirtir) [reply CHANGELOG.md:34]: Also, should we update the release date for 12.0?

Comment (hanefi) [reply CHANGELOG.md]: I had a chat with Aykut, and merged the non-transactional fix with the related s919fe72.

I also set the date to today in 037ea72
